### ISSUE TITLE
Implicit cast to bool gives unexpected behavior for SArray

### ISSUE THREAD

**Issue by znation** [dato-code/SFrame#88]
When I do the following:
```
x = gl.SArray([1,None,0])
y = gl.SArray([0,0,1])
print x and y
```

I expected that (like many other operators), the `and` operator would be overloaded to do an element-wise and. The result I get actually has the expected type and expected number of rows, but _not_ the result I expected. I get back `y`:
```
[0, 0, 1]
```

This is because apparently Python likes to implicitly-cast-to-bool for comparison purposes, and return the original type from the expression (JS does this as well, so I should've realized that's what was happening). The equivalent expression to `x and y`, without implicit type coercion, is actually `y if bool(x) else x`.

[Explicit is better than implicit](https://www.python.org/dev/peps/pep-0020/) probably applies here. Why Python originally allowed this implicit casting is beyond me.

@hoytak suggests that we adopt Numpy behavior here and error on implicit cast to bool, and I'm inclined to agree. I assume `TypeError`?

**Comment by ylow** [comment]
This as it turns out, is kind of a critical bug fix since it raised several bugs in our unit tests.
Specifically ```bool(sa)``` is currently defined as ```len(sa) != 0``` which seems sensible.

However, this results in an interesting issue in unittests where ```self.assertEquals(sa, [... values...])``` may result in a pass **even if the values do not match**. This is because internally, self.assertEquals performs ```bool(sa == [...values ...])``` which is true even if the values do not match.

**Comment by ylow** [comment]
Fixed in #274 . ```bool(sa)``` now raises an exception.
